File: js/test/htdocs/spec/test_checksum.js

describe('Checksum', function() {

    it('Collisions', function() {
		for (length = 10; length <= 1000; length += 200) {
            testCollisions(length);
        }
    });
    
    it('No input data', function() {
		const bridge = new JsMidiBridge(null, null);
		
		expect(bridge.getChecksum(new Uint8Array())).toEqual(new Uint8Array([0, 0, 0]));
    });
    
    
});

function testCollisions(length) {
	const bridge = new JsMidiBridge(null, null);
	
	const data = generateData(length);
	expect(data.length).toBe(length);
	
	const checksum = bridge.getChecksum(data);
	
	expect(checksum).toBeInstanceOf(Uint8Array);
	expect(checksum.length).toBe(3);
	
	// Change bits
	for (i = 2; i < length; i += 5) {
	    const changed_list = data.slice();
	    if (changed_list[i] < 127) {
	        changed_list[i] += 1;
	    } else {
	        changed_list[i] -= 1;
        }
	
	    const checksum_changed =  bridge.getChecksum(changed_list);
	
		expect(checksum).not.toEqual(checksum_changed);
	}
}

function generateData(length) {
	ret = [];
	for (i = 0; i < length; ++i) {
		ret.push(i % 256);
	}
	return new Uint8Array(ret);
}

/*



   

    def test_no_data(self):
       

 */

File: python/test/pymidibridge/test_checksum.py
import sys
import unittest
from unittest.mock import patch   # Necessary workaround! Needs to be separated.

from .mocks import *

with patch.dict(sys.modules, {
    "time": MockTime
}):
    from lib.pymidibridge.PyMidiBridge import *


class TestChecksum(unittest.TestCase):

    def test_collisions(self):
        for length in range(10, 1000, 200):
            self._test_collisions(length)

    def _test_collisions(self, length):
        bridge = PyMidiBridge(None, None)

        data = self._generate_data(length)
        self.assertEqual(len(data), length)

        checksum = bridge._get_checksum(data)

        msg = "Data: " + repr(data) + "; Checksum: " + repr(checksum)

        self.assertEqual(len(checksum), 3, msg)
        
        # Change bits
        for i in range(2, length, 5):
            changed_list = list(data)
            if changed_list[i] < 127:
                changed_list[i] += 1
            else:
                changed_list[i] -= 1

            checksum_changed =  bridge._get_checksum(bytes(changed_list))

            self.assertNotEqual(checksum, checksum_changed)


    def _generate_data(self, length):
        return bytes([i % 256 for i in range(length)])


    def test_no_data(self):
        bridge = PyMidiBridge(None, None)

        self.assertEqual(bridge._get_checksum(bytes([])), b'\x00\x00\x00')
        self.assertEqual(bridge._get_checksum(None), b'\x00\x00\x00')
        self.assertEqual(bridge._get_checksum(0), b'\x00\x00\x00')

